### PR TITLE
add store_eager_result setting so eager tasks can store result on the backend

### LINKED ISSUE #6476: `publish_result` is False when tasks are eager and and we DON'T want to ignore results
#### Mandatory Debugging Information

- [ ] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

#### Related Issues

- #5398 
- https://github.com/celery/django-celery-results/issues/49

#### Possible Duplicates

- None

### Python Packages

Latest, not released django-celery-results
https://github.com/celery/django-celery-results/pull/172/commits/b4acbf24ea3521ecfdbf821ece7ca62e49ff2765
Currently published version doesn't support Celery 5.x, but all versions are affected

## Minimally Reproducible Test Case

Source of the issue
https://github.com/celery/celery/blob/2a6c7cfe3b1283961887bf1cb3f5aa6c8aa70820/celery/app/trace.py#L323

```python
import pytest

@pytest.mark.parametrize(
    "eager,ignore_result,publish_result",
    [
        (True, True, False),
        (True, False, True),  # fails
        (False, True, False),
        (False, False, True),
    ]

)
def test_publish_result_current(eager, ignore_result, publish_result):
    actual_publish_result = not eager and not ignore_result
    assert publish_result == actual_publish_result


@pytest.mark.parametrize(
    "eager,ignore_result,publish_result",
    [
        (True, True, False),
        (True, False, True),
        (False, True, False),
        (False, False, True),
    ]

)
def test_publish_result_desired(eager, ignore_result, publish_result):
    actual_publish_result = not ignore_result
    assert publish_result == actual_publish_result
```

# Expected Behavior

``always_eager`` exists so we can use Celery in unit tests. I want to execute a task eagerly and later on check result from the backend. 

# Actual Behavior
When task is eager and it's set NOT to ignore the result, the result will be ignored.

### PR DESCRIPTION
**Pull request by tomwojcik**
Currently, if you use **eager** tasks and set that you **do not want to ignore the result**, it won't be saved on the backend.
There's literally no possible way to do that. 

This PR adds a setting that will allow doing just that.

Imo that's a bug. Fixing it by allowing the task result to be saved when ignore result is false might break a lot of already existing projects so adding a new setting is the safest option.

Fixes #6476

### CODE REVIEW

**Review by auvipy — COMMENTED** [review]
looks good to me. did you check this manually that it's working?

**Comment by tomwojcik** [comment]
> looks good to me. did you check this manually that it's working?

Not yet, I plan to test it today against my project. Will get back to you with the confirmation.

**Comment by tomwojcik** [comment]
@auvipy I confirm it is working! :)

**Comment by tomwojcik** [comment]
Though there is one issue, eager tasks do not store FAILURE results. Is that a blocker?


https://github.com/celery/celery/blob/29eda054555fa95c83210e5e6bc3e839c80bcd3b/celery/app/trace.py#L162

**Comment by auvipy** [comment]
> Though there is one issue, eager tasks do not store FAILURE results. ~Is that a blocker?~
> 
> Added.
> 
> https://github.com/celery/celery/blob/29eda054555fa95c83210e5e6bc3e839c80bcd3b/celery/app/trace.py#L162

thanks for checking. since it's a old issue so we had confusion in mind, so manual testing was needed

**Comment by tomwojcik** [comment]
> thanks for checking. since it's a old issue so we had confusion in mind, so manual testing was needed

Sure. Just to make it clear, I am currently using this version in my project and I find it to be working flawlessly. I'm not sure what _old _issue__ you are talking about.

Please let me know if you need something else from me.

**Comment by auvipy** [comment]
> > thanks for checking. since it's a old issue so we had confusion in mind, so manual testing was needed
> 
> Sure. Just to make it clear, I am currently using this version in my project and I find it to be working flawlessly. I'm not sure what _old _issue__ you are talking about.
> 
> Please let me know if you need something else from me.

I meant this is a very old issue so was bit cautious. I think we are good. waiting for some response from other maintainers

**Comment by matusvalo** [comment]
As long as I understand the main motivation of this PR is to support testing. Maybe we can update also this documentation:

https://docs.celeryproject.org/en/latest/userguide/testing.html

**Comment by tomwojcik** [comment]
> As long as I understand the main motivation of this PR is to support testing. Maybe we can update also this documentation:
> 
> https://docs.celeryproject.org/en/latest/userguide/testing.html

Yes, makes sense. Will do.

**Comment by tomwojcik** [comment]
Sorry it took so long. For the past 3 weeks I use it in my project, everything is fine. I believe tests exhaust all new cases and there will be no regression as it is a new setting. I think there's nothing more to be added in the docs either.

Please let me know if I can do anything else.

**Review by matusvalo — DISMISSED** [review]
LGTM for me.

**Review by auvipy — COMMENTED** [review]
IMHO we should now focus on 5.1 new features to master

File: docs/userguide/configuration.rst
Comment (tomwojcik) [reply docs/userguide/configuration.rst:432]: right, added

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
merging this as 5.0 release almost 6 months back. we should combine 5.0.6 & 5.1.0 release to 5.1.0 this time. hope everyone is ok with that reasoning. I will try to help to write the release notes so that releasing would be easier for whoever wants to release them.

**Review by thedrow — COMMENTED** [review]
Thank you. Sorry I didn't get to reviewing this before it got merge.
Looks great!

**Pull request merged**
